PhotoMap/Features/Map/ViewModels/MapViewModel.swift

```diff
@@ -4,94 +4,134 @@
-
         locationService.status
             .filter { $0 == .denied }
             .map { _ in () }
             .subscribe(coordinator.disableLocationSubject)
             .store(in: cancelBag)
+
+        loadUserPhotosSubject
+            .debounce(for: 0.5, scheduler: RunLoop.main)
+            .flatMap { [unowned self] visibleRect in
+                self.firestoreService.getPhotos(by: visibleRect)
+            }
+            .sink(receiveCompletion: сompletionHandler,
+                  receiveValue: { [weak self] photos in self?.photos = photos })
+            .store(in: cancelBag)
+
+        tapMapViewGestureSubject
+            .sink(receiveValue: { [weak self] _ in self?.enableDiscoveryMode()})
+            .store(in: cancelBag)
+    }
+
+    private func enableDiscoveryMode() {
+        userTrackingMode = .none
+    }
+
+    private func сompletionHandler(completion: Subscribers.Completion<FirestoreError>) {
+        switch completion {
+        case .failure(let error):
+            coordinator.errorAlertSubject.send(error)
+        case .finished:
+            return
+        }
+    }
+}
+
+extension MapViewModel: MKMapViewDelegate {
+
+    func mapViewDidChangeVisibleRegion(_ mapView: MKMapView) {
+        loadUserPhotosSubject.send(mapView.visibleMapRect)
+    }
+
+    func mapView(_ mapView: MKMapView, didSelect view: MKAnnotationView) {
+        guard let photoView = view as? PhotoMarkerView,
+              let photoAnnotation = view.annotation as? PhotoAnnotation else { return }
+
+        guard let url = photoAnnotation.imageUrl else { return }
+
+        if !photoView.isLoadImage {
+            firestoreService.downloadImage(by: url)
+                .sink(receiveCompletion: self.сompletionHandler,
+                      receiveValue: { image in
+                        guard let image = image else { return }
+                        photoView.loadImageSubject.send(image)
+                      })
+                .store(in: cancelBag)
+        }
+    }
+
+    func mapView(_ mapView: MKMapView,
+                 annotationView view: MKAnnotationView,
+                 calloutAccessoryControlTapped control: UIControl) {
+        guard let photoAnnotation = view.annotation as? PhotoAnnotation else { return }
+        // Open FullPhoto screen
     }
 
-    private func switchFollowDiscoveryMode(disableFolowMode: Bool = false) {
-        isFollowModeOn = disableFolowMode ? !disableFolowMode : !isFollowModeOn
-        modeButtonCollor = isFollowModeOn ? Asset.followModeColor.color : Asset.discoverModeColor.color
+    func mapView(_ mapView: MKMapView, didChange mode: MKUserTrackingMode, animated: Bool) {
+        switch mode {
+        case .follow:
+            modeButtonTintColor = Asset.followModeColor.color
+        case .none:
+            modeButtonTintColor = Asset.discoverModeColor.color
+        case .followWithHeading:
+            enableDiscoveryMode()
+        @unknown default:
+        fatalError()
```

Comment:
Will be a good surprise for users if Apple introduce new `MKUserTrackingMode` later :)
Just `break`